CIP-1694/README.md

```diff
@@ -0,0 +1,664 @@
+* a role - Constitutional Committee, DRep, or SPO;
+* a key-hash (which will be verified to have the role above);
+* a URL for any metadata that is relevant to the vote;
+* a hash of the contents of this URL;
+* a yes/no/abstain vote.
+
+Note that "abstained" votes are not included in the "active voting stake".
+Note also that a vote to abstain is different than abstaining from voting
+(the former is visible on chain, the latter is not).
+To avoid confusion, we will only use the word "abstain" from this point onward to mean a vote
+to abstain.
+
+The key hash will trigger the appropriate signature check on the transaction body according to the existing `UTxOW` ledger rule.
+
+Votes can be cast multiple times for each governance action by a single key hash. Correctly submitted votes override any older votes for the same key hash and role.
+As soon as a governance action is ratified, voting ends. Any future votes are not considered or recorded (whether they are yes, no or abstain).
+
+#### Governance State
+
+When a governance action is successfully submitted to the chain, its progress will be tracked by the ledger state.
+In particular, the following will be tracked:
+
+* the governance action ID;
+* the epoch that the action expires;
+* the deposit amount;
+* the rewards address that will receive the deposit when it is returned;
+* the total yes/no/abstain votes of the Constitutional Committee for this action;
+* the total yes/no/abstain votes of the DReps for this action;
+* the total yes/no/abstain votes of the SPOs  for this action.
+
+#### Stale Votes
+
+The votes from the DReps and the SPOs may become meaningless after crossing an epoch boundary since they may become unregistered. Therefore, all unregistered votes are cancelled before any new votes are considered.
+
+#### Changes to the Stake Snapshot
+
+Since the stake snapshot changes at each epoch boundary, a new current voting tally must be calculated for each governance action based on the votes that have been cast before any new votes are counted.  This avoids the same stake being used two or more times.  _This new tally may result in immediate ratification of a governance action._
+
+### Delegated Representatives (DReps)
+
+> **Warning**
+> The design of the DReps is still subject to change and should not be conflated with Project Catalyst DReps.
+
+Existing stake keys will be given new responsibilities. They will be able to delegate their stake to DReps. DRep Registration will mimic the existing stake delegation mechanisms (via certificates).
+
+The following new types of voting certificates will be added:
+
+* DRep registration certificate. Including:
+  * DRep ID: hash of verification key
+  * a URL for any metadata that is relevant to the DRep
+  * a hash of the contents of this URL
+
+* DRep retirement certificate. Including:
+  * DRep ID
+  * epoch number after which the DRep will retire
+
+* Vote delegation certificate. Including:
+  * stake credential
+  * DRep ID
```

Comment:
Thank you for your thoughtful and well-prepared proposal! It is clear that a great deal of effort and consideration went into its creation.

It seems that the vote delegation certificate will need to have a one-to-many structure (one stake credential to multiple DRep IDs), rather than the current one-to-one design. This is because it would be difficult for wallets to implement voting for users if a user wants to delegate her funds into different DReps.

For example, consider a user who has 10 ADA and delegates to a stake pool `P0`. If the same user wants to split her voting power evenly between two different DReps, `D0` and `D1`, the current design would require the user (or the wallet) to create a new stake key, send 5 ADA to the new stake key address, create a new stake delegation certificate that delegates to P0, and create vote delegation certificates to D0 and D1. This process would be complex for any wallet to implement, and it would become even more complicated if the user wanted to delegate to multiple pools and DReps at the same time. In addition, when a user wants to receive funds, she would need to carefully choose the receiving address based on the intended voting power delegation, which could be a poor user experience.

As a voter, I would prefer a user interface that allows me to easily control the percentage of voting power delegated to each DRep without creating new stake keys. Using the example above, I would want the voting power for D0 to be 50%, and the same for D1. To achieve this, I suggest modifying the structure of the vote delegation certificate to include a list of DRep IDs and a corresponding list of percentages indicating the voting power that each DRep in the list should receive. The sum of the percentage list should be between 0 and 100. For instance, the vote delegation certificate would look like this for the example described above:

```
* Vote delegation certificate
  * stake credential
  * [D0, D1]
  * [50, 50]
```

Happy to discuss if you have better ideas or don't agree with the proposed design. As a side note, the same design could be used to support multi-pool stake delegation as well, but that is out of the scope of this CIP.

Reply:
> It seems that the vote delegation certificate will need to have a one-to-many structure (one stake credential to multiple DRep IDs), rather than the current one-to-one design. This is because it would be difficult for wallets to implement voting for users if a user wants to delegate her funds into different DReps.

This is a great topic, thank you for bringing it up @cffls . Our previous head architect, @johnalanwoods, actually wrote a [decision memo](https://github.com/input-output-hk/cardano-ledger/blob/ea48d4815946c4f35239eb41815fe06cd6ffec2e/docs/proxy-keys/proxy-key-sols.tex) about exactly this topic in February of last year. Embarrassingly, my team's usual way of hosting PDFs isn't working at the moment, but here is a [google drive link](https://drive.google.com/file/d/1Gg2jefHmRHSEaoLTilSj1YGxDZ-onWuj/view?usp=sharing) to the PDF. The memo recommends the "wallet based solution", not the "ledger based solution".

Note that at least one wallet that I am aware of, namely [Typhon](https://typhonwallet.io), has implemented the "wallet based solution". This could, perhaps, make for a good case study.

Reply:
Thanks Jared for providing the link to previous work! I went through the memo, but it seems like the ledger solution described in the doc is to use proxy certificate, which is different from the enhanced solution I proposed above. Therefore, the disadvantages outlined in the memo for proxy certificates do not seem to apply here.

In addition, the memo only considers the staking scenario. With voting combined, the complexity of implementing wallet-based solution will increase significantly, especially in the case where a user's staking delegation and voting delegation have different distribution. For example, a user wants to delegate to three pools with a distribution of 10%, 20%, and 70%, while delegate her voting power to five DReps with 7%, 9%, 12%, 15%, and 43% distribution. It seems like creating many staking keys is inevitable with the wallet solution, which will bring confusions to end users. The wallet implement is also a non-trivial problem. Even if possible, what would the default receiving address a wallet should recommend user to use? I strongly recommend to reconsider this design, and involve major wallet teams, e.g. Typhon, Eternl, Nami, Flint, Lace, into the discussion and listen to their feedbacks as well.

Reply:
> the ledger solution described in the doc is to use proxy certificate

That's definitely a fair point. Many proposals have been made over the past few years for "partial delegation at the protocol layer", and indeed what you suggested is not the one in the memo. At least drawback 6, privacy, still applies.

> In addition, the memo only considers the staking scenario. 

One of the big advantages of the current DRep design is that it can re-use the vast majority of our existing infrastructure and design. For the first version of Voltaire, we wanted to piggy-back off of the staking scenario as much as possible.

> the complexity of implementing wallet-based solution will increase significantly

It doesn't seem that complicated to me, you just need `m×n` many stake credentials, where `m` is the number of stake pools that you want to delegate to and `n` is the number of DReps that you want to delegate to. Obviously this would grow quickly if `m` or `n` got big, but as long as we are only talking about a few pools and DReps it should be fine.

> which will bring confusions to end users

It's not clear to me why the wallet cannot hide this from the user, in the way Typhon already does today.

> Even if possible, what would the default receiving address a wallet should recommend user to use?

I'm definitely not a UI designer, but off the top of my head, maybe two drop downs where you select one of the pools you are delegated to and one of the DReps you are delegated to, and out displays an address?

> I strongly recommend to reconsider this design, and involve major wallet teams, e.g. Typhon, Eternl, Nami, Flint, Lace, into the discussion and listen to their feedbacks as well.

Absolutely, I'm happy to hear from everyone. That said, it's not clear to me that we want to add partial delegation (a potentially big topic) to an already complex CIP.

Reply:
Thanks for the quick response! Below are my thoughts.

> That's definitely a fair point. Many proposals have been made over the past few years for "partial delegation at the protocol layer", and indeed what you suggested is not the one in the memo. At least drawback 6, privacy, still applies.

If privacy is crucial, we can still use wallet solution right? The enhanced ledger solution doesn't enforce user to put all DReps into one certificate.

> One of the big advantages of the current DRep design is that it can re-use the vast majority of our existing infrastructure and design. For the first version of Voltaire, we wanted to piggy-back off of the staking scenario as much as possible.

Understood, but it would still require additional work from wallet side. At the moment, only one wallet has implemented multi-pool delegation though. 

> It doesn't seem that complicated to me, you just need m×n many stake credentials, where m is the number of stake pools that you want to delegate to and n is the number of DReps that you want to delegate to. Obviously this would grow quickly if m or n got big, but as long as we are only talking about a few pools and DReps it should be fine.

The cost of creating mxn UTxOs + delegation certificate is much larger than just publishing one certificate. Also, if a user decides to add one more pool or DRep, rebalancing will cost much more than publishing a new certificate.

> I'm definitely not a UI designer, but off the top of my head, maybe two drop downs where you select one of the pools you are delegated to and one of the DReps you are delegated to, and out displays an address?

I can imagine many users are not interested in selecting which of the pool and DReps should receive an incoming staking/voting power. All they care is that the fund are received correctly, and the distribution of staking and voting powers stay the same. If I am receiving a large fund to an address delegated to a particular pool and DRep, resulting a significant change in the distribution of my staking/voting power, I have to create another transaction that rebalance my wallet after, which is definitely not a good user experience. Ideally, receiving funds to a wallet should be an orthogonal event to the distribution of the staking and voting power.

> Absolutely, I'm happy to hear from everyone. That said, it's not clear to me that we want to add partial delegation (a potentially big topic) to an already complex CIP.

Thanks! I understood that this is a big topic, but I am not convinced we should simply use the existing delegation method simply because it is a big topic. I can't see strong arguments about the disadvantages of the proposed design other than the fact that it could be a big topic and involve more works.

Reply:
> The cost of creating mxn UTxOs + delegation certificate is much larger than just publishing one certificate. Also, if a user decides to add one more pool or DRep, rebalancing will cost much more than publishing a new certificate.

At 44 lovelace per byte (the current mainnet value) the difference is pretty negligible. I totally agree, however, that re-balancing is the worst part about the wallet solution.

> If privacy is crucial, we can still use wallet solution right? The enhanced ledger solution doesn't enforce user to put all DReps into one certificate.

Sure, but what I'm afraid of is some other group wanting something more like proxy keys, or one of the other proposals, and the protocol becomes more complex than it needs to be.

> I am not convinced we should simply use the existing delegation method simply because it is a big topic. I can't see strong arguments about the disadvantages of the proposed design other than the fact that it could be a big topic and involve more works.

I'm not trying to end the discussion, I'm just trying to move it. My stance is this:

* If we make changes to support partial delegation at the protocol level, let's do it in the exactly same way for _both_ distributions, block production and voting.
* This topic is orthogonal and complex enough to merit its own CIP. In fact, I think that CIP-36 actually already covers all of these ideas (though it uses transaction metadata since the original target was catalyst).

Note that we do not have a lot of breathing room with the performance of the calculations involving the stake credentials (since there are millions of them). For example, just last month we added [this optimization](https://github.com/input-output-hk/cardano-ledger/blob/583a13e80336666979ad0c2a98375987bee950f3/docs/adr/2022-12-12_007-optimize-ledger-view.md) in order to mitigate things like [this](https://github.com/input-output-hk/cardano-node/issues/4421). And now we are about to add a new collapsing of the per-stake-credential distribution. We just need to be careful and not add too many things all at once.

Another reason not to couple partial delegation with CIP-1694 is that we are still in the middle of the [on-disc](https://github.com/input-output-hk/ouroboros-network/issues?q=is%3Aissue+is%3Aopen+label%3A%22UTxO-HD+%F0%9F%93%92%F0%9F%92%BD%22) project, which aims to move all of the "large" data structures in the ledger from memory to disc. The first goal is to move the UTxO to disc, but after that we will be moving the stake distributions. This is another reason to proceed with caution before adding any new complexity.

Reply:
Thank you, Jared, for taking the time to respond. Your explanation makes perfect sense. My original goal for starting this thread was to bring attention to this potential issue, in case it hadn't been discussed or addressed before. I'd be happy to continue discussing this topic in other CIPs. Please feel free to resolve this comment. Thanks!